### PR TITLE
Fix the `context_subtree()` exporter cache and its span-processor leaks

### PR DESCRIPTION
**Pull request by dsfaccini**
> This pull request was posted by Codex Desktop using gpt-5.6-sol on behalf of David.

### Why we make these changes

`_add_context_span_exporter()` cached exporters by provider `id()` and returned cache hits without validating them. Provider shutdown could therefore reuse a stopped exporter, while a recycled `id()` could reuse an exporter attached to a different provider. Both cases made `context_subtree()` silently produce an empty `SpanTree`.

The cache now pairs each exporter with the provider it is attached to, matches providers by identity, rejects stopped exporters, and weakly evicts entries where possible. The same transaction prevents concurrent callers from attaching duplicate processors and keeps the cache key and attachment target on the same resolved provider.

This is separate from [#7070](https://github.com/pydantic/pydantic-ai/pull/7070), which finalizes span trees when task bodies raise, and [#7603](https://github.com/pydantic/pydantic-ai/pull/7603), which fixed the same empty-tree symptom when OpenTelemetry context leaked.

### New public surface

none

### User-visible behavior

```diff
pydantic_evals/otel/context_subtree.py :: context_subtree()
  └─ pydantic_evals/otel/_context_in_memory_span_exporter.py :: _add_context_span_exporter()
-   └─ return the exporter cached under `id(provider)`
-       └─ stopped or wrong-provider exporter → empty `SpanTree`
+   └─ validate provider identity and exporter lifecycle under one lock
+       └─ attached active exporter → populated `SpanTree`
```

<details><summary>Cache transaction and lock shape</summary>

- One `Lock` covers lookup, creation, attachment, and cache publication, so concurrent first callers cannot attach duplicate processors.
- The entry is published only after `add_span_processor()` succeeds, so a failed attachment does not poison the cache and the next call retries.
- Weak-referenceable providers evict their entries when collected. Providers that cannot be weakly referenced are pinned, avoiding a fresh leaked processor on every call.
- Logfire's wrapped provider is resolved once and used for validation, cache identity, and attachment, so a concurrent provider swap cannot separate the entry from its processor.
- The non-reentrant lock deliberately rejects re-entry through `add_span_processor()` rather than allowing the duplicate attachment the lock prevents. Supported SDK and Logfire providers do not re-enter this call path.

</details>

### Verification

The regression tests cover [shutdown recovery and provider identity/reuse](https://github.com/pydantic/pydantic-ai/blob/a1b2078ec59f6adea61be935ac59af6be8815122/tests/evals/test_otel.py#L1041-L1154), [Logfire provider swaps and failed-attachment retries](https://github.com/pydantic/pydantic-ai/blob/a1b2078ec59f6adea61be935ac59af6be8815122/tests/evals/test_otel.py#L1157-L1226), and [weak eviction, non-weak-referenceable providers, recycled identities, and concurrent registration](https://github.com/pydantic/pydantic-ai/blob/a1b2078ec59f6adea61be935ac59af6be8815122/tests/evals/test_otel.py#L1228-L1378). The targeted test file and 100%-coverage gate pass.

### What changes for existing users

Nothing unless they hit one of the broken cache or concurrent-registration paths above.

One further case, listed for completeness rather than because it is reachable: where the global tracer provider is a Logfire proxy wrapping something that is not an SDK `TracerProvider`, `context_subtree()` now yields a `SpanTreeRecordingError` (which `EvaluatorContext.span_tree` raises) instead of an empty `SpanTree`. The `add_span_processor` check moved from the proxy to the provider it wraps, and Logfire's proxy silently no-ops that call when the wrapped provider is not an SDK one, which is what previously produced the empty tree. `logfire.configure()` installs an SDK provider before the proxy becomes global and never reverts it, so this state is not producible through Logfire's own API.

### Checklist

- [ ] Any **AI generated code** has been reviewed line-by-line by the human PR author, who stands by it.
- [x] No **breaking changes** in accordance with the [version policy](https://github.com/pydantic/pydantic-ai/blob/main/docs/version-policy.md).
- [x] Any permitted **compatibility impact** has a label, warning, migration, release note, and exact API-check waiver.
- [x] **PR title** is fit for the [release changelog](https://github.com/pydantic/pydantic-ai/releases).

### CODE REVIEW

File: pydantic_evals/pydantic_evals/otel/_context_in_memory_span_exporter.py [hunk 1]
```diff
@@ -162,8 +157,30 @@ def _add_context_span_exporter() -> _ContextInMemorySpanExporter | SpanTreeRecor
...
+    # provider is what identifies the set of span processors; the proxy itself is also unhashable.
+    if isinstance(tracer_provider, LogfireProxyTracerProvider):
+        cache_key = tracer_provider.provider
+    else:
+        cache_key = tracer_provider
+
+    # A provider that is unhashable or cannot be weakly referenced cannot key the cache. That is not fatal:
+    # recording still works, it just attaches a fresh exporter per call instead of reusing one.
+    try:
+        cached_exporter = _context_in_memory_providers.get(cache_key)
```
[the remark this answers is not in the document]
Comment (dsfaccini) [reply pydantic_evals/pydantic_evals/otel/_context_in_memory_span_exporter.py]: Valid, and reproduced: two distinct `ValueEqualTracerProvider` instances that compare equal shared one exporter, and the second had zero processors attached, so it recorded nothing.

Root cause was mine — I keyed a `WeakKeyDictionary` by the provider object, which matches on equality. Fixed in a0e7c1cad by pairing each entry with a weak reference to its provider, keyed by `id()` and matched by identity, so neither hashability nor equality is required. Pinned by `test_context_span_exporter_not_shared_between_equal_providers`.

File: pydantic_evals/pydantic_evals/otel/_context_in_memory_span_exporter.py [hunk 2]
```diff
@@ -162,8 +157,30 @@ def _add_context_span_exporter() -> _ContextInMemorySpanExporter | SpanTreeRecor
...
+    # A provider keeps its identity across `shutdown()`, which stops the exporter attached to it. A stopped
+    # exporter silently drops every span it is handed, so it has to be replaced rather than reused. The dead
+    # processor stays attached, so a provider shut down repeatedly without being replaced accumulates one per
+    # shutdown -- bounded in practice because `logfire.configure()` allocates a new provider each time.
+    if cached_exporter is not None and not cached_exporter._stopped:  # pyright: ignore[reportPrivateUsage]
+        return cached_exporter
+
     exporter = _ContextInMemorySpanExporter()
-    _context_in_memory_providers[cache_id] = exporter
+    with suppress(TypeError):
```
[the remark this answers is not in the document]
Comment (dsfaccini) [reply pydantic_evals/pydantic_evals/otel/_context_in_memory_span_exporter.py]: Valid, and reproduced: four calls against an uncacheable provider attached four processors and four distinct exporters, and only the newest is ever cleared, so the rest retain spans permanently.

Fixed in a0e7c1cad. Switching to identity keying removed the cause for the realistic shape — unhashable providers (any `@dataclass`, which is what logfire's own proxy is) are now cached normally rather than falling into the uncached path. Only a non-weak-referenceable provider stays uncached, since its entry could never be invalidated; that one is covered by `test_context_subtree_provider_that_cannot_be_cached`.

The same review also led me to lock the check-then-act: racing callers each attached a processor and only the winner's exporter stayed reachable, which is the same leak by a different route. `test_context_span_exporter_attached_once_under_concurrency` holds the window open and fails without the lock.

**Comment by dsfaccini** [comment]
> This comment was posted by Claude Code using claude-opus-5 on behalf of David.

🔍 Local review suite (automated, pre-merge)

| Wave | Status | Blocking | Required (do) | Warnings (defer/skip) |
|---|---|---|---|---|
| 1 · code-correctness | ✅ closed | 0 | 0 | 8 |
| 2 · tests/cassettes/docs | ✅ closed | 0 | 0 | 3 |
| 3 · cross-cutting/siblings | ✅ closed | 0 | 0 | 8 |

**Verdict:** ✅ PASS — was 🔧 9 required; all 9 resolved in `a1b2078e` · 9 applied
Reviewed at 2026-09-01 · diff `sha256:08917544`

<details>
<summary>✅ 9 required action(s) resolved</summary>

_Automated review — flagged by Claude, not yet confirmed by a maintainer._
_The 9 required actions this suite flagged at `6fedc625`, all resolved in `a1b2078e`; CI green on that head, coverage gate held._

**1. The re-entrancy comment omitted that the pre-PR code handled re-entry** · `_context_in_memory_span_exporter.py` · `_add_context_span_exporter`
`6fedc625`'s unreachability claim is correct — verified in the installed sources, not just by tests: `opentelemetry-sdk` 1.44.0's `SynchronousMultiSpanProcessor._lock` is taken by `add_span_processor` alone, and logfire 4.40.0's proxy is off the path since `.provider` is resolved first. But the removed path *was* load-bearing for one shape: merge base and `88e15cc8` both complete with 1 processor where `6fedc625` hangs forever. The comment now says a re-entrant custom provider deadlocks where the implementation this replaced returned, and that provisional publication under an `RLock` does hold both properties — it was built and dropped, not impossible.

**2. The concurrency test could not see the defect class it guards** · `tests/evals/test_otel.py` · `test_context_span_exporter_attached_once_under_concurrency`
Its only assertion was `len(processors) == 1`, which stays green while 4 exporters are constructed and 3 of 4 callers hold an unattached one. Now also pins the construction count and that the attached and cached exporter are the one built.

**3. The shutdown caveat named one spec clause and one path** · `_context_in_memory_span_exporter.py`, `test_context_subtree_records_after_plain_sdk_provider_shutdown`
`SpanProcessor.Shutdown` is a second, independent `SHOULD`; because `add_span_processor` appends to the composite processor `shutdown()` stopped, it starves the Logfire path too. Both the comment and the docstring now say so.

**4. `AttachmentFailed(BaseException)` → `Exception`** · `tests/evals/test_otel.py`

**5. Block comment no longer bounded by a trailing test name**, and no longer claims the tests only read the cache — one seeds it · `tests/evals/test_otel.py`

**6. `from weakref import ref` → `import weakref`** (5 sites) · `_context_in_memory_span_exporter.py`

**7. Renamed to `test_context_span_exporter_pins_a_provider_that_cannot_be_weakly_referenced`**, matching the prefix its siblings use for cache-state assertions · `tests/evals/test_otel.py`

**8. `type(tracer_provider).__qualname__` → `type(provider).__qualname__`** — the guard had moved to the resolved provider; the message had not · `_context_in_memory_span_exporter.py`

**9. PR body now names the Logfire-proxy behavior change** (empty `SpanTree` → `SpanTreeRecordingError`), and its Verification links were retargeted from `6fedc625` to `a1b2078e` after the line ranges shifted.

</details>

<details>
<summary>⚠️ 19 informational (defer/skip)</summary>

_Automated review — flagged by Claude, not yet confirmed by a maintainer._

Verified-and-skipped, recorded so they are not re-raised: the process-global lock held across a third-party call (measured 0.85s cross-provider stall with a 1s attach; a tuple append with real providers, and no lock-order inversion — the SDK's `_lock` is taken by `add_span_processor` alone); dead processors accumulating on a repeatedly-shut-down retained provider (inherent to the fix, bounded under Logfire); the lock not being fork-safe (reproduced, but nothing in the four packages forks, and three sibling module-level locks carry no fork handler); test residue in the module-global cache (measured: exactly 2 entries at session end, order-independent across three orderings, neither mis-servable); `_add_context_span_exporter` at 82 lines (every extraction moves the attach outside the lock that owns the invariant); attaching past Logfire's proxy; the permanent strong pin for non-weak-referenceable providers.

**Refused on the evidence** — raised as findings, not acted on: widening the `isinstance` to `(ProxyTracerProvider, LogfireProxyTracerProvider)` (the state cannot arise — `_initialize` calls `set_provider(SDKTracerProvider)` before `trace.set_tracer_provider`, once, and never reverts, so the widened arm would need a test for an unreachable state under `fail_under = 100`); and "the Logfire skipif costs the `pydantic-evals` CI lane its coverage of this file" (that lane resolves no `opentelemetry-sdk` at all, so the module is unimportable there regardless — no fixable gap to track).

**Rescoped to pre-existing:** the blocking-`threading.Lock`-inside-a-Temporal-workflow concern reached via the `OnlineEvaluation` capability. The merge base already acquires a `threading.Lock` twice per `context_subtree()` round trip on that same path and already mutates a module-global cache; this diff widens the critical section rather than introducing the class. No durable-exec module references `pydantic_evals`, so there is no verified reach to file against.

</details>

File: pydantic_evals/pydantic_evals/otel/_context_in_memory_span_exporter.py [hunk 3]
```diff
@@ -162,8 +170,48 @@ def _add_context_span_exporter() -> _ContextInMemorySpanExporter | SpanTreeRecor
...
+            # where leaving it uncached would attach a fresh span processor on every call and leave every
+            # orphaned exporter collecting spans that nothing ever clears. A pinned provider can never be
+            # freed, so its `id()` can never be recycled either.
+            stored = provider
+
+        processor = SimpleSpanProcessor(exporter)
+        # Cached only once the attach has succeeded: an entry written first would claim an attachment that
+        # never happened, and every later call would return an exporter no provider feeds -- a silently empty
+        # `SpanTree`, which is the failure this cache exists to avoid.
+        provider.add_span_processor(processor)  # pyright: ignore[reportAttributeAccessIssue, reportUnknownMemberType]
```
[the remark this answers is not in the document]
Comment (dsfaccini) [reply pydantic_evals/pydantic_evals/otel/_context_in_memory_span_exporter.py:229]: > This reply was posted by Claude Code using claude-opus-5 on behalf of David.

Valid and reproduced — but both suggested remedies are worse than the deadlock, so I've documented the constraint instead of "fixing" it.

`RLock` doesn't make this reentrancy-safe. The re-entrant call finds no cache entry yet (the entry is written only after the attach succeeds) and attaches a **second** span processor — the exact leak this lock was added to prevent. Moving the attach outside the lock reopens the race that motivated it. There's no third option: you can't both serialise the attach and permit the attach to re-enter.

The trigger also doesn't exist in any real provider: `opentelemetry-sdk`'s `add_span_processor` appends under its own `_lock`, and logfire's `ProxyTracerProvider` forwards under `self.lock`. Neither calls back into `pydantic_evals`.

<details><summary>Measured, both arms</summary>

A provider whose `add_span_processor` re-enters `context_subtree()` once:

```
Lock         -> DEADLOCK (timed out)
RLock        -> completed, processors attached = 2
```

</details>

The comment above the lock now states the constraint and why neither remedy is taken.

File: pydantic_evals/pydantic_evals/otel/_context_in_memory_span_exporter.py [hunk 4]
```diff
@@ -128,23 +126,33 @@ def force_flush(self, timeout_millis: int = 30000) -> bool:  # pragma: no cover
...
-_context_in_memory_providers: WeakValueDictionary[int, _ContextInMemorySpanExporter] = WeakValueDictionary()
+# Caching the exporter per provider keeps `context_subtree()` from attaching another span processor on every
+# call: a long-lived provider would otherwise accumulate one per evaluation. Each entry pairs the exporter with the
+# provider it is attached to, keyed by `id()` and matched by identity. Identity is what matters here, not equality:
+# a provider need not be hashable, and two distinct providers that compare equal each need their own exporter, since
+# an exporter only ever receives spans from the provider it was attached to. The provider is held weakly wherever it
+# can be, which is what makes an `id()` key safe -- `id()`s are recycled, so an entry outliving its provider must be
+# rejected rather than handed to whatever was later allocated at the same address.
+_context_in_memory_providers: dict[int, tuple[ref[TracerProvider] | TracerProvider, _ContextInMemorySpanExporter]] = {}
+_context_in_memory_providers_lock = threading.RLock()
```
[the remark this answers is not in the document]
Comment (dsfaccini) [reply pydantic_evals/pydantic_evals/otel/_context_in_memory_span_exporter.py]: > This reply was posted by Claude Code using claude-opus-5 on behalf of David.

**The sentence you quoted is the defect, not the justification.** Replacing it is why the rest of the diff exists.

On the merge base (`286a2ac`), one ordinary run of `tests/evals/test_online.py` served **214 of 248** cache lookups an exporter that had already been shut down. A stopped exporter returns `FAILURE` from `export()` and drops every span silently. Being handed a dead exporter is the steady state within a couple of tests of any `logfire.shutdown()`, not a synthetic edge case.

Being precise about that number: those evaluations emit no spans, so nothing was lost *there* — it measures how often the broken state arises, not lost data. Where spans do flow, the loss is real and silent.

**You're right about one part, and I removed it** in 6fedc6250. `opentelemetry-sdk`'s `TracerProvider.add_span_processor` and logfire's `ProxyTracerProvider.add_span_processor` both only append under their own lock — neither calls back into user code, so no supported provider can re-enter `context_subtree()`. The provisional publication, the rollback, the `RLock` and its subprocess-isolated test defended a case that cannot occur. Back to a plain `Lock` publishing after a successful attach: −55 lines, no coverage change.

**Narrowing the rest to identity-keying alone doesn't survive contact with the tests.** Seven remaining tests fail against the merge-base implementation, each pinning a different block:

<details><summary>The seven, and what each one pins</summary>

Reverting only `_context_in_memory_span_exporter.py` to `286a2ac` and running `tests/evals/test_otel.py`:

| Test | Block it pins |
|---|---|
| `test_context_subtree_records_after_tracer_provider_shutdown` | stopped-exporter rejection, logfire path |
| `test_context_subtree_records_after_plain_sdk_provider_shutdown` | stopped-exporter rejection, plain SDK path |
| `test_context_span_exporter_attaches_to_the_logfire_provider_it_caches` | resolving the wrapped provider once, so key and attach target match |
| `test_context_span_exporter_retries_after_an_attachment_failure` | publishing the entry only after the attach succeeds |
| `test_context_subtree_provider_that_cannot_be_weakly_referenced` | pinning a provider `weakref.ref()` rejects, instead of leaking a processor per call |
| `test_context_span_exporter_refused_when_its_provider_was_recycled` | identity match, which is what makes an `id()` key sound |
| `test_context_span_exporter_attached_once_under_concurrency` | the lock spanning attach, not just the cache write |

Identity-keying alone fixes the third and sixth. It leaves the shutdown arm — the 214 — untouched.

</details>

Two concessions: there is no linked issue, so the above is a measurement rather than a field report; and the shutdown recovery does rest on `opentelemetry-sdk` letting a newly attached processor receive spans after `shutdown()`, which the spec does not require. The plain-SDK test's docstring carries that caveat.

Agreed on coordinating landing order with #7070.
